File: problem_solving/math/cpp/9_is_palindrome.cpp
/*
https://leetcode.com/problems/palindrome-number/

note you dont have to fully reverse the number.
better option it to reverse the second half...  till the rev_number > original num..
*/
#include<iostream>

bool is_number_palindrome(int number) {
    if ((number < 0) || (number %10 == 0 && x !=0)) {
        return false;
    }
    int rev_number = 0;
    while (rev_number <= number ) {
        rev_number = rev_number*10 + number % 10;
        if(rev_number == number) {
            return true;
        }
        number /= 10;
        if(rev_number == number) {
            return true;
        }
    }
    return false;
}



int main() {
    int num = 12321;
    bool result = is_number_palindrome(num);
    std::cout << "result: " << result << std::endl;

    num = 123216;
    result = is_number_palindrome(num);
    std::cout << "result: " << result << std::endl;

    num = 11;
    result = is_number_palindrome(num);
    std::cout << "result: " << result << std::endl;
    return 0;
}

File: problem_solving/math/py/9_is_palindrome.py
'''
https://leetcode.com/problems/palindrome-number/
'''

def is_palindrome_number(x:int):
    if x == 0:      #mistake - missed the validations here
        return True
    if x < 0 or x %10 == 0:
        return False

    reverse_number = 0
    original_num = x
    i = 0
    while x:
        r = x % 10
        x = x//10
        reverse_number = reverse_number*10  + r     #mistake ... did + instead of x
        i += 1
    print("rev: " + str(original_num) + " => " + str(reverse_number))
    return reverse_number == original_num


x=123
result=is_palindrome_number(x)
print("result:"+ str(x)+" => "+str(result))


x=121
result=is_palindrome_number(x)
print("result:"+ str(x)+" => "+str(result))

x=-121
result=is_palindrome_number(x)
print("result:"+ str(x)+" => "+str(result))

x=10
result=is_palindrome_number(x)
print("result:"+ str(x)+" => "+str(result))




